Create Search Page

shayakrach opened the pull request:
Filtering jobs by the search cell in the Navigation bar by title.
Only registered (HR's and students) and log-in users can use
the search feature. This search feature increases user satisfaction
by improving the usability, accessibility, and efficiency of user
interaction with the website.

Close #61 

I recognized a typo that I made in a previous PR.
This typo will be fixed in another PR that closes issue #71.

AshkenYariv reviewed (approved):
Looks good!

assafg95 commented:
Looks good to me too

ormergi commented on templates/base.html:
```diff
@@ -38,8 +38,8 @@
-            <form class="d-flex">
-              <input class="form-control me-2" type="search" placeholder="Search" aria-label="Search">
+            <form class="d-flex" action="{% url 'jobs_search' %}" method="get">
+              <input name="q" class="form-control me-2" type="search" placeholder="Search" aria-label="Search...">
```
What is `q`? please rename it to more informative name.

ormergi replied:
If its not something that related to Django, please rename `q` to more informative (e.g: search_query).

ormergi reviewed (commented):
first commit:
Why this commit is necessary for this PR?
If it can be done on another PR please move it.

ormergi reviewed (changes requested):
second commit:
Please add tests for the new business logic (**at least**: search actually return what you expect, redirection to search page works).
Add them to this commit and not on another one.

ormergi commented:
@shayakrach  ping

shayakrach commented:
> first commit: Why this commit is necessary for this PR? If it can be done on another PR please move it.

It is not.
Moved to #77.

shayakrach commented:
> @shayakrach ping

Hi @ormergi,

This PR is in WIP status right now.

When it will be ready I will post it on Slack,

shayakrach committed:
Create Search Page

Filtering jobs by the search cell in the Navigation bar by title.
Only registered (HR's and students) and log-in users can use
the search feature. This search feature increases user satisfaction
by improving the usability, accessibility, and efficiency of user
interaction with the website.

shayakrach commented:
> first commit: Why this commit is necessary for this PR? If it can be done on another PR please move it.

Already moved.

shayakrach commented:
> second commit: Please add tests for the new business logic (**at least**: search actually return what you expect, redirection to search page works). Add them to this commit and not on another one.

Fixed @ormergi

ormergi commented on conftest.py:
```diff
@@ -97,5 +97,33 @@ def job_2(user_HR):
                          is_active=True,
                          marked_count=0,
-                         apply_link=DEFAULT_APPLY_LINK)
+                         apply_link=DEFAULT_APPLY_LINK+'2')
```
Why this change is here?

ormergi commented on conftest.py:
```diff
@@ -97,5 +97,33 @@ def job_2(user_HR):
+
+
+@pytest.fixture
+def job_SoftwareEngineer(user_HR):
```
This function name does not follow [python function name convention ](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)
and its not consistent with the reset of the project.
Please change it and other similar function names.

Also, it seems that you use the new fixtures only for the search page tests, please move them to the same file.

ormergi commented on conftest.py:
```diff
@@ -97,5 +97,33 @@ def job_2(user_HR):
+
+
+@pytest.fixture
+def job_DataEngineer(user_HR):
```
Same

ormergi commented on jobs/tests/test_jobs_view.py:
```diff
@@ -30,6 +30,7 @@ def test_unauthenticated_user_enter_register(client):
 
 
 # Should move move another file - Good for Demo 3
+# Fix the typo here - issue #71
```
There is already an issue open for this, there is no need to add a comment as well especially on  non related PR.

ormergi commented on jobs/tests/test_search_jobs_view.py:
```diff
@@ -0,0 +1,46 @@
+
+
+@pytest.mark.django_db
+def test_search_page_filtering_by_title_successfully_by_any_partial_title(client, user_HR, job_DataEngineer):
```
I would remove `..search_page..` part from all the fixture names,
their names already long and it could be understood by the file name.
Something like `test_search_by_partial_job_title`

ormergi commented on jobs/tests/test_search_jobs_view.py:
```diff
@@ -0,0 +1,46 @@
+
+@pytest.mark.django_db
+def test_search_page_filtering_by_title_successfully_by_any_partial_title(client, user_HR, job_DataEngineer):
+    assert Job.objects.count() == 1
```
I think this assertion should be part of `job_DataEngineer` fixture

ormergi commented on jobs/tests/test_search_jobs_view.py:
```diff
@@ -0,0 +1,46 @@
+    assert Job.objects.count() == 1
+
+    client.force_login(user_HR)
+    title = job_DataEngineer.title + 'e'
```
This one deserves its own fixture instead of manipulation one that exists.

ormergi reviewed (changes requested):
Hi Shay, 

I started to review this PR starting for reading its description then reviewed the the tests in order to understand how it works.
From what I understand this feature is targeted only for HR users,
if this is not the case I would expect to see various users types on tests.
Also I would expect to see at least tree elements on the job page when you test the search feature.

The PR description should also include motivation for the feature (how its improving the app e.g: UX improvement), who can use it (HR users only?), same goes for the commit message.

shayakrach commented:
> 



> Hi Shay,
> 
> I started to review this PR starting for reading its description then reviewed the the tests in order to understand how it works. From what I understand this feature is targeted only for HR users, if this is not the case I would expect to see various users types on tests. Also I would expect to see at least tree elements on the job page when you test the search feature.
> 
> The PR description should also include motivation for the feature (how its improving the app e.g: UX improvement), who can use it (HR users only?), same goes for the commit message.



> Hi Shay,
> 
> I started to review this PR starting for reading its description then reviewed the the tests in order to understand how it works. From what I understand this feature is targeted only for HR users, if this is not the case I would expect to see various users types on tests. Also I would expect to see at least tree elements on the job page when you test the search feature.
> 
> The PR description should also include motivation for the feature (how its improving the app e.g: UX improvement), who can use it (HR users only?), same goes for the commit message.

I made all the changes you asked for @ormergi .
I think it is better now.

assafg95 reviewed (approved):
Everything Looks Great!

ormergi commented on jobs/tests/test_search_jobs_view.py:
```diff
@@ -0,0 +1,170 @@
+    software_engineer_jobs = []
+    for i in range(NUM_OF_JOBS):
+        software_engineer_job = Job.create_job(
+                                publisher=user_HR, title='Software Engineer',
```
nit: move title to new line.

ormergi commented on jobs/tests/test_search_jobs_view.py:
```diff
@@ -0,0 +1,170 @@
+    data_engineer_jobs = []
+    for i in range(NUM_OF_JOBS):
+        data_engineer_job = Job.create_job(
+                            publisher=user_HR, title='Data Engineer',
```
nit: move title to new line.

ormergi reviewed (changes requested):
Hi Shay,

Please change the vars and function names according to python naming convention, they arr also apply for acronyms (HR -> hr).

Regarding tests fixture names, try to rephrase them to be more like natural language, for example `test_search_by_partial_job_title_by_HR` -> `test_hr_user_job_search_by_title`

Note about tests, in general its a good practice to treat tests as the documentation of your project,
it allows new comers to understand the flow, how it works and how to use it (in case it is a library).

shayakrach commented:
> Hi Shay,
> 
> Please change the vars and function names according to python naming convention, they arr also apply for acronyms (HR -> hr).
> 
> Regarding tests fixture names, try to rephrase them to be more like natural language, for example `test_search_by_partial_job_title_by_HR` -> `test_hr_user_job_search_by_title`
> 
> Note about tests, in general its a good practice to treat tests as the documentation of your project, it allows new comers to understand the flow, how it works and how to use it (in case it is a library).

Fixed @ormergi

ormergi reviewed (approved):
Thanks!
